sdks/go/pkg/beam/core/runtime/exec/translate.go

```diff
@@ -411,11 +413,11 @@ func (b *builder) makePCollection(id string) (*PCollection, error) {
 }
 
 func (b *builder) newPCollectionNode(id string, out Node) (*PCollection, error) {
-	ec, _, err := b.makeCoderForPCollection(id)
+	ec, wc, err := b.makeCoderForPCollection(id)
 	if err != nil {
 		return nil, err
 	}
-	u := &PCollection{UID: b.idgen.New(), Out: out, PColID: id, Coder: ec, Seed: rand.Int63()}
+	u := &PCollection{UID: b.idgen.New(), Out: out, PColID: id, Coder: ec, WindowCoder: wc, Seed: rand.Int63(), dataSampler: b.dataSampler}
```

Comment:
I'll note that there's at least one location where we *do not* want to have Data Sampling behavior, and that's for Large Iterables.

The `mayFixDataSourceCoder` function has 2 cases where we convert the coder to a CoGBK handler. We should nil the DataSource's PCollection dataSampler in those cases.

The reasoning is values grouped by key are typically large, so we probably shouldn't sample them. We can always evaluate this more specifically later if there's a problem or customer demand. 

Also, since the iterator Types are anonymous functions being passed in, I don't think it's going to print anything meaningful, if at all. (It'll at best, encode the key, and then the "iterators" would print their function pointers, if it prints anything for the function pointers at all).

I guess that's really the question: What gets sampled currently for GBKs in your tests?

Reply:
Scratch that: Due to how the SDK is implemented, we currently never sample anything from the DataSource with your code. We simply count and size all the elements since we have the raw bytes already.

https://github.com/apache/beam/blob/5c6955ad00b2fdef412b4b7325b94ac78db1dc20/sdks/go/pkg/beam/core/runtime/exec/datasource.go#L244

I'd say we don't block on fixing this issue, but we should add a TODO around there, with a note about validating CoGBKs and similar types with the sampling implementation. We'd also just want to copy the bytes "raw" for a sampled windowed value element element (the PCollection sampling logic, which we'd want to then also co-opt) instead of decoding an element, and then re-encoding it. That might avoid some questions around Large Iterables, or large GBKs too.

Reply:
1. Correct. That's the intent.
2. Probably not, since except for right at the DataSource, we don't have bytes, we have elements. We must encode them to get the bytes.

WRT 2. The implementation I suggested in the 2 paths for Sampling available, vs not, avoids duplicate encodings, so it's about as good as we're getting. We also don't sample that frequently (outside of the streaming case with tiny bundle sizes).